[ML][FEATURE] SPARK-5566: RegEx Tokenizer

aborsu985 opened the pull request:
Added a Regex based tokenizer for ml.
Currently the regex is fixed but if I could add a regex type paramater to the paramMap,
changing the tokenizer regex could be a parameter used in the crossValidation.
Also I wonder what would be the best way to add a stop word list.

Commit:
RegExTokenizer

A more complex tokenizer that extracts tokens based on a regex. It also allows
to turn lowerCasing on and off, adding a minimum token length and a list of
stop words to exclude.

Commit:
RegEx Tokenizer

A more complex tokenizer that extracts tokens based on a regex. It also allows
to turn lowerCasing on and off, adding a minimum token length and a list of
stop words to exclude.

AmplabJenkins commented:
Can one of the admins verify this patch?

hhbyyh commented:
Just FYI, there's some similar code in LDAExample.

aborsu985 commented:
This is not meant to be a standalone tokenizer but rather part of a pipeline.
In that aim it has parameters that can be made to vary in order to decide which are best.
I did tthis in response to this issue.
https://issues.apache.org/jira/browse/SPARK-5566

hhbyyh commented:
Cool, just didn't saw that in the PR description. 
And if possible, apart from the stopwords, an extra parameter representing the vocab will be handy. In some cases, the vocabulary has been previously settled. Just my thought.

aborsu985 commented:
Do you mean restricting tokens to a predestined set of words?

mengxr commented:
@aborsu985 Please follow the steps in https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark to prepare a PR.

If this is called `RegexTokenizer`, should the regex used to be configurable?

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
RegexTokenizer

Regex and stopwords parameters are now part of the parametergrid

aborsu985 commented:
@mengxr 
I changed the title to be more specific about the change and enabled the regex to be configurable (as well as the stopwords). There is an issue on the Jira about this (https://issues.apache.org/jira/browse/SPARK-5566) and the code passes the tests when I run it on my machine. The change is quite small, adds no dependencies and should be useful to anyone who wants to validate a NLP model using pipelines.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+  def setLowercase(value: Boolean) = set(lowerCase, value)
+  def getLowercase: Boolean = get(lowerCase)
+
+  val minLength = new IntParam(this, 
```
Add doc and update code style. What's the case when we match a token with regex but its length is zero? Should we control it in the regex, e.g., `\d{5,}`?

Btw, it is not intuitive that the min value is excluded. Could we remove "excluded" and set the default value to `1`? And it might be better to call it `minTokenLength`, if we want to keep it.

aborsu985 replied:
I removed excluded as it is indeed unusual and set the default value to 1 which is standard

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+
+/**
+ * :: AlphaComponent ::
+ * A regex based tokenizer that extracts tokens using a regex.
```
From the doc, it is not clear to me whether I should specify groups to extract in the regex or it repeatedly matches the entire regex.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+@AlphaComponent
+class RegexTokenizer extends UnaryTransformer[String, Seq[String], RegexTokenizer] {
+
+  val lowerCase = new BooleanParam(this, 
```
Add doc with `@group param` to group methods in the generated html doc, e.g., https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/param/sharedParams.scala#L65

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+class RegexTokenizer extends UnaryTransformer[String, Seq[String], RegexTokenizer] {
+
+  val lowerCase = new BooleanParam(this, 
+      "lowerCase", 
```
We don't chop down args in a method call.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+
+  val lowerCase = new BooleanParam(this, 
+      "lowerCase", 
+      "enable case folding to lower case", 
```
It is not clear to me whether this is applied before regex matching or after.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+  def setMinLength(value: Int) = set(minLength, value)
+  def getMinLength: Int = get(minLength)
+
+  val regEx = new Param(this, 
```
`regEx` -> `regex`. The latter is more common. Or we can use `tokenRegex` to be more specific.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+  val regEx = new Param(this, 
+      "regEx", 
+      "RegEx used for tokenizing", 
+      Some("\\p{L}+|[^\\p{L}\\s]+".r))
```
Using `Regex` as param type is not Java/Python friendly. We should use plain string.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+  def setRegex(value: scala.util.matching.Regex) = set(regEx, value)
+  def getRegex: scala.util.matching.Regex = get(regEx)
+
+  val stopWords = new Param(this, 
```
I would leave this function to a stopword filter.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+  def setStopWords(value: Array[String]) = set(stopWords, value)
+  def getStopWords: Array[String] = get(stopWords)
+
+
```
remove extra blank line

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+
+
+  override protected def createTransformFunc(paramMap: ParamMap): String => Seq[String] = { x =>
+
```
remove this line too

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+
+  override protected def createTransformFunc(paramMap: ParamMap): String => Seq[String] = { x =>
+
+    var string = x
```
Try to use `val` if possible:

~~~
val str = if (paramMap(lowerCase)) x.toLowerCase else s
~~~

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+    if (paramMap(lowerCase)) {
+      string = string.toLowerCase
+    }
+    var tokens = (paramMap(regEx) findAllIn string).toList
```
Do not use infix (https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-InfixMethods).

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+    }
+    var tokens = (paramMap(regEx) findAllIn string).toList
+    
+    if(paramMap(minLength) > 0){
```
space before `(paramMap` and after `> 0)`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,66 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+      tokens = tokens.filter(_.length > paramMap(minLength))
+    }
+    if(paramMap(stopWords).length > 0){
+      tokens = tokens.filter(!paramMap(stopWords).contains(_))
```
`stopWords` is an `Array[String]`and hence `contains` has linear time complexity. We should convert it into a set first. As I mentioned before, I think the stop words could be handled separately.

mengxr commented:
@aborsu985 I made a pass on the code. Besides my inline comments, please add a unit test. It would be better if you can also add a Java unit test. Thanks!

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Removed stopwords parameters and updated doc

Still need to add unit test.

aborsu985 commented:
Changed minimum token length to 1 and removed the excluded bit.
Added matching param which allows to switch from matching regex to splitting regex.
Reduced length of Param doc so that lines are smaller than 100 chars
RegEx is now regex and type String
StopWord param has been removed as it should be it's own feature modification.

Need to add param doc and unit test

Commit:
Added doc to RegexTokenizer

Added @groups {param, setParam, getParam} documentation to RegexTokenizer

mengxr commented:
@aborsu985 Sorry for the delay! On the high level, I'm a little concerned with exposing too many parameters in the first version. NLTK's regex tokenizer (http://www.nltk.org/_modules/nltk/tokenize/regexp.html) only has three parameters: `pattern`, `gaps`, and `discard_empty`. There are trade-offs between having a full featured implementation and maintenance cost. For example, if we put `lowerCase` as a parameter here, what if a user just want to convert something into lowercase letters? Should there be a separate implementation of `LowerCase` as a transformer? Then we will have function overlaps. Maybe we can remove `lowerCase` in this PR.

For other parameters, `minTokenLength` looks good to me, which is a superset of NLTK's `discard_empty`. The name `matching` doesn't tell you what happens if it is `false`. I don't think NLTK's `gaps` is doing a better job either. But given the fact that there has been many NLTK users, using `gaps` may be a better choice. Feel free to propose any names used in popular NLP packages.

mengxr commented:
ok to test

mengxr commented:
add to whitelist

SparkQA commented:
[Test build #28544 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28544/consoleFull) for   PR 4504 at commit [`d3ef6d3`](https://github.com/apache/spark/commit/d3ef6d32ee4e9da8feb2db5291763dbb8b53b1ea).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28544 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28544/consoleFull) for   PR 4504 at commit [`d3ef6d3`](https://github.com/apache/spark/commit/d3ef6d32ee4e9da8feb2db5291763dbb8b53b1ea).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RegexTokenizer extends UnaryTransformer[String, Seq[String], RegexTokenizer] `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28544/
Test PASSed.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Changed matching to gaps and removed case folding

SparkQA commented:
[Test build #28727 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28727/consoleFull) for   PR 4504 at commit [`132b00b`](https://github.com/apache/spark/commit/132b00bd236121f90e7c3aa9270a64d3ede4417d).
 * This patch merges cleanly.

aborsu985 commented:
@mengxr I do not think that LowerCase warrants a transformer but rather it could be incorporated into a larger string to vector transformer that changes a text into a vector just like the [CountVectorizer](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html#sklearn.feature_extraction.text.CountVectorizer) used in scikit learn. 
I agree that it is not necessary to have it in the tokenizer though so I removed it.
I'm working on adding unit tests but I'm having trouble creating a dataFrame to test it with. Any help would be welcome.

Commit:
Changed regex to pattern

As we changed matching to gaps to be consistent with nltk's RegexTokenizer,
it is also more consistent to use pattern instead of regex.

SparkQA commented:
[Test build #28728 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28728/consoleFull) for   PR 4504 at commit [`cd6642e`](https://github.com/apache/spark/commit/cd6642e9810135c6600cdc0bf209c5f7c264949b).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28727 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28727/consoleFull) for   PR 4504 at commit [`132b00b`](https://github.com/apache/spark/commit/132b00bd236121f90e7c3aa9270a64d3ede4417d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RegexTokenizer extends UnaryTransformer[String, Seq[String], RegexTokenizer] `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28727/
Test PASSed.

SparkQA commented:
[Test build #28728 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28728/consoleFull) for   PR 4504 at commit [`cd6642e`](https://github.com/apache/spark/commit/cd6642e9810135c6600cdc0bf209c5f7c264949b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RegexTokenizer extends UnaryTransformer[String, Seq[String], RegexTokenizer] `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28728/
Test PASSed.

mengxr commented:
You can use https://github.com/apache/spark/blob/master/mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala as a template for unit tests. In `beforeAll` construct a DataFrame with a string column.

Commit:
Added unit tests in scala

Also changed RegexTokenizer so it extends Tokenizer Class instead of UnaryTransformer
It might be interesting to create a Tokenizer trait that could be used by all tokenizers

SparkQA commented:
[Test build #28798 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28798/consoleFull) for   PR 4504 at commit [`e262bac`](https://github.com/apache/spark/commit/e262bacc5fc3cc780466f49200cdf2466ce8e563).
 * This patch merges cleanly.

aborsu985 commented:
Thank you for the tip, I'll look into the java tests next week when I have some time.
But in the meantime. I changed the RegexTokenizer to extend from Tokenizer instead of UnaryTransformer. This seemed like a good idea at the time as I could then test the two tokenizers with one function. But this seems to cause some problems in java. I'm looking into it.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,67 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+ * lowercase = false, minTokenLength = 1
+ */
+@AlphaComponent
+class RegexTokenizer extends Tokenizer {
```
This is not necessary at this time. It doesn't affect any user code because we don't expect users to treat tokenizers in a generic way.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,67 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+  def getMinTokenLength: Int = get(minTokenLength)
+
+  /**
+   * param sets regex as splitting on gaps(true) or matching tokens (false)
```
space before `(true)`

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,94 @@
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+
```
remove extra empty line

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,94 @@
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+
+case class TextData(rawText : String,wantedTokens: Seq[String])
```
space after `,` and put an empty line after.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,94 @@
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    sqlContext = new SQLContext(sc)
```
initialize `dataset` here to avoid duplicate code. If they are not the same in tests, please remove dataset from TokenizerSuite and define local variables in each test.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,94 @@
+    testTokenizer(oldTokenizer,dataset)
+  }
+
+  def testTokenizer(t: Tokenizer,dataset: DataFrame){
```
space before `{`

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  def testTokenizer(t: Tokenizer,dataset: DataFrame){
+  	t.transform(dataset)
```
Please configure your editor to not use tabs.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,94 @@
+      .select("tokens","wantedTokens")
+      .collect().foreach{ 
+        case Row(tokens: Seq[String], wantedTokens: Seq[String]) =>
+          assert(tokens.length == wantedTokens.length)
```
`assert(tokens === wantedTokens)` should work.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,94 @@
+          assert(tokens.length == wantedTokens.length)
+          tokens.zip(wantedTokens).foreach(x => assert(x._1 == x._2))
+        case _ => 
+          println()
```
remove `println` and throw an exception with the content of the Row.

SparkQA commented:
[Test build #28798 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28798/consoleFull) for   PR 4504 at commit [`e262bac`](https://github.com/apache/spark/commit/e262bacc5fc3cc780466f49200cdf2466ce8e563).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RegexTokenizer extends Tokenizer `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28798/
Test PASSed.

mengxr commented:
@aborsu985 Please check the code style and make sure you follow https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide

Commit:
Modified the pattern Param so it is compiled when given to the Tokenizer

pattern is set and get as a string but stored as a compiled regex this
prevents having to recompile it everytime the transform function is called

Commit:
Added Java unit test for RegexTokenizer

This unit tests only tests that the getters and setters work properly from java and tries
one transform. Other tests have already been carried out in the TokenizerSuite.scala test.

Commit:
Style corrections

SparkQA commented:
[Test build #28862 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28862/consoleFull) for   PR 4504 at commit [`6a85982`](https://github.com/apache/spark/commit/6a859820513050a41b46294d741e99f57c999a56).
 * This patch merges cleanly.

aborsu985 commented:
Sorry my commit was a bit hasty. Any automated style checkers to recommend?

SparkQA commented:
[Test build #28862 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28862/consoleFull) for   PR 4504 at commit [`6a85982`](https://github.com/apache/spark/commit/6a859820513050a41b46294d741e99f57c999a56).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RegexTokenizer extends UnaryTransformer[String, Seq[String], RegexTokenizer] `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28862/
Test PASSed.

sagacifyTestUser committed:
Merge remote-tracking branch 'upstream/master'

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,67 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+  /**
+   * param for minimum token length, default is one to avoid returning empty strings
+   * @group param
+   */
```
Please append `: IntParam` to `minTokenLength`. See SPARK-6428. Please also update other method declarations.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala:
```diff
@@ -39,3 +39,67 @@ class Tokenizer extends UnaryTransformer[String, Seq[String], Tokenizer] {
+  val minTokenLength = new IntParam(this, "minLength", "minimum token length", Some(1))
+
+  /** @group setParam */
+  def setMinTokenLength(value: Int) = set(minTokenLength, value)
```
`: this.type`

mengxr commented on mllib/src/test/java/org/apache/spark/ml/feature/JavaTokenizerSuite.java:
```diff
@@ -0,0 +1,73 @@
+      .setGaps(true)
+      .setMinTokenLength(0);
+
+    List<String> t = Arrays.asList(
```
Unit tests should be minimal. Pulling in JSON RDD is not necessary. Please check how `LabeledDocument` is defined and used in Java:

https://github.com/apache/spark/blob/4a17eedb16343413e5b6f8bb58c6da8952ee7ab6/examples/src/main/scala/org/apache/spark/examples/ml/SimpleTextClassificationPipeline.scala#L29

https://github.com/apache/spark/blob/4a17eedb16343413e5b6f8bb58c6da8952ee7ab6/examples/src/main/java/org/apache/spark/examples/ml/JavaCrossValidatorExample.java#L74

aborsu985 replied:
That was actually what I tried to do in the first place.
After solving the problem of passing a java.util.List to a Scala object, the RDD created with the code you showed me results in a empty dataframe being created and the test fails as the column with the name demanded by the tokenizer is not found. After spending far too much time on this I changed to using the json to RDD.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,103 @@
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+case class TextData(rawText : String, wantedTokens: Seq[String])
```
insert an empty line between class declarations.
remove space before `:`.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,103 @@
+      .setOutputCol("tokens")
+
+    var dataset = sqlContext.createDataFrame(
+      sc.parallelize(List(
```
minor: `Seq` is used more often than `List` in the MLlib codebase.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,103 @@
+
+    var dataset = sqlContext.createDataFrame(
+      sc.parallelize(List(
+        TextData("Test for tokenization.",List("Test","for","tokenization",".")),
```
space after `,` (please also check this in other places.)

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,103 @@
+  def testTokenizer(t: Tokenizer,dataset: DataFrame): Unit = {
+    t.transform(dataset)
+      .select("tokens","wantedTokens")
+      .collect().foreach{ 
```
space before `{` (please also check this in other places)

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/TokenizerSuite.scala:
```diff
@@ -0,0 +1,103 @@
+        case Row(tokens: Seq[Any], wantedTokens: Seq[Any]) =>
+          assert(tokens === wantedTokens)
+        case e =>
+          throw new SparkException(s"Row $e should contain only tokens and wantedTokens columns")
```
`SparkException` should happen on workers. Since data is already collected, we can use `fail("...")`. For this test, maybe the following is sufficient:

~~~scala
.collect()
.foreach { case Row(actual, expected) =>
  assert(actual === expected)
}
~~~

mengxr commented:
I don't know a formatter that can do everything correctly. I use intellij and with the default Scala code style (except indent 2). I need to manually adjust the indentation while chopping down the args.

sagacifyTestUser committed:
Merge remote-tracking branch 'upstream/master'

sagacifyTestUser committed:
Added return type to public functions

Plus some cosmetic changes.

SparkQA commented:
[Test build #28990 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28990/consoleFull) for   PR 4504 at commit [`148126f`](https://github.com/apache/spark/commit/148126fc0af7682ea951632a29033a0f9d464811).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28990 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28990/consoleFull) for   PR 4504 at commit [`148126f`](https://github.com/apache/spark/commit/148126fc0af7682ea951632a29033a0f9d464811).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RegexTokenizer extends UnaryTransformer[String, Seq[String], RegexTokenizer] `
  * `class NaiveBayesModel(Saveable, Loader):`
  * `class SqlParser extends AbstractSparkSQLParser with DataTypeParser `
  * `case class CombineSum(child: Expression) extends AggregateExpression `
  * `case class CombineSumFunction(expr: Expression, base: AggregateExpression)`
  * `protected[sql] class DataTypeException(message: String) extends Exception(message)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28990/
Test PASSed.

mengxr committed:
change pattern to a StringParameter; update tests

sagacifyTestUser committed:
Merge remote-tracking branch 'upstream/master'

SparkQA commented:
[Test build #29059 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29059/consoleFull) for   PR 4504 at commit [`2338da5`](https://github.com/apache/spark/commit/2338da5410cb6846dfcf8288f2ceb0376aa8995d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29059 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29059/consoleFull) for   PR 4504 at commit [`2338da5`](https://github.com/apache/spark/commit/2338da5410cb6846dfcf8288f2ceb0376aa8995d).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ExecutorCacheTaskLocation(override val host: String, executorId: String)`
  * `class RegexTokenizer extends UnaryTransformer[String, Seq[String], RegexTokenizer] `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29059/
Test PASSed.

mengxr committed:
Merge remote-tracking branch 'apache/master' into SPARK-5566

mengxr committed:
update test

mengxr committed:
Merge branch 'aborsu985-master' into SPARK-5566

mengxr committed:
remove tabs

mengxr commented:
@aborsu985 I sent you a PR with some updates at https://github.com/aborsu985/spark/pull/1. Please merge the current master and check the diff. Thanks!

sagacifyTestUser committed:
Merge remote-tracking branch 'upstream/master'

SparkQA commented:
[Test build #29153 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29153/consoleFull) for   PR 4504 at commit [`5f09434`](https://github.com/apache/spark/commit/5f09434f7b60fbff4fab2899256a96155dc75981).
 * This patch merges cleanly.

sagacifyTestUser committed:
Merge branch 'SPARK-5566' of git://github.com/mengxr/spark into mengxr-SPARK-5566

sagacifyTestUser committed:
Merge branch 'mengxr-SPARK-5566'

SparkQA commented:
[Test build #29154 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29154/consoleFull) for   PR 4504 at commit [`716d257`](https://github.com/apache/spark/commit/716d2579f0f0829753cdd5882d22a5e497486a3e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29153 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29153/consoleFull) for   PR 4504 at commit [`5f09434`](https://github.com/apache/spark/commit/5f09434f7b60fbff4fab2899256a96155dc75981).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RegexTokenizer extends UnaryTransformer[String, Seq[String], RegexTokenizer] `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29153/
Test PASSed.

SparkQA commented:
[Test build #29154 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29154/consoleFull) for   PR 4504 at commit [`716d257`](https://github.com/apache/spark/commit/716d2579f0f0829753cdd5882d22a5e497486a3e).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RegexTokenizer extends UnaryTransformer[String, Seq[String], RegexTokenizer] `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29154/
Test PASSed.

aborsu985 commented:
@mengxr Thank you for your help with the Java unit tests. As you may have guessed, I'm new to both Scala and Java and I was drowning in it.

mengxr commented:
LGTM. Merged into master. Thanks for contributing!
